File: android/src/main/java/org/maplibre/reactnative/components/mapview/LayerSourceInfo.java
package org.maplibre.reactnative.components.mapview;

import org.maplibre.android.style.layers.CircleLayer;
import org.maplibre.android.style.layers.FillExtrusionLayer;
import org.maplibre.android.style.layers.FillLayer;
import org.maplibre.android.style.layers.HeatmapLayer;
import org.maplibre.android.style.layers.HillshadeLayer;
import org.maplibre.android.style.layers.Layer;
import org.maplibre.android.style.layers.LineLayer;
import org.maplibre.android.style.layers.RasterLayer;
import org.maplibre.android.style.layers.SymbolLayer;

import javax.annotation.Nullable;

class LayerSourceInfo {
    final String sourceId;

    @Nullable
    final String sourceLayerId;

    LayerSourceInfo(Layer layer) {
        if (layer instanceof CircleLayer) {
            CircleLayer symbolLayer = (CircleLayer) layer;
            sourceId = symbolLayer.getSourceId();
            sourceLayerId = symbolLayer.getSourceLayer();
        } else if (layer instanceof FillExtrusionLayer) {
            FillExtrusionLayer fillExtrusionLayer = (FillExtrusionLayer)layer;
            sourceId = fillExtrusionLayer.getSourceId();
            sourceLayerId = fillExtrusionLayer.getSourceLayer();
        } else if (layer instanceof FillLayer) {
            FillLayer fillLayer = (FillLayer)layer;
            sourceId = fillLayer.getSourceId();
            sourceLayerId = fillLayer.getSourceLayer();
        } else if (layer instanceof HeatmapLayer) {
            HeatmapLayer heatmapLayer = (HeatmapLayer)layer;
            sourceId = heatmapLayer.getSourceId();
            sourceLayerId = heatmapLayer.getSourceLayer();
        } else if (layer instanceof HillshadeLayer) {
            HillshadeLayer hillshadeLayer = (HillshadeLayer)layer;
            sourceId = hillshadeLayer.getSourceId();
            sourceLayerId = null;
        } else if (layer instanceof LineLayer) {
            LineLayer lineLayer = (LineLayer)layer;
            sourceId = lineLayer.getSourceId();
            sourceLayerId = lineLayer.getSourceLayer();
        } else if (layer instanceof RasterLayer) {
            RasterLayer rasterLayer = (RasterLayer) layer;
            sourceId = rasterLayer.getSourceId();
            sourceLayerId = null;
        } else if (layer instanceof SymbolLayer) {
            SymbolLayer symbolLayer = (SymbolLayer)layer;
            sourceId = symbolLayer.getSourceId();
            sourceLayerId = symbolLayer.getSourceLayer();
        } else {
            sourceId = "";
            sourceLayerId = null;
        }
    }
}

File: android/src/main/java/org/maplibre/reactnative/components/mapview/helpers/LayerSourceInfo.kt
package org.maplibre.reactnative.components.mapview.helpers

import org.maplibre.android.style.layers.CircleLayer
import org.maplibre.android.style.layers.FillExtrusionLayer
import org.maplibre.android.style.layers.FillLayer
import org.maplibre.android.style.layers.HeatmapLayer
import org.maplibre.android.style.layers.HillshadeLayer
import org.maplibre.android.style.layers.Layer
import org.maplibre.android.style.layers.LineLayer
import org.maplibre.android.style.layers.RasterLayer
import org.maplibre.android.style.layers.SymbolLayer

internal class LayerSourceInfo(
    layer: Layer?,
) {
    val sourceId: String

    val sourceLayerId: String?

    init {
        when (layer) {
            is CircleLayer -> {
                val symbolLayer = layer
                sourceId = symbolLayer.getSourceId()
                sourceLayerId = symbolLayer.getSourceLayer()
            }

            is FillExtrusionLayer -> {
                val fillExtrusionLayer = layer
                sourceId = fillExtrusionLayer.getSourceId()
                sourceLayerId = fillExtrusionLayer.getSourceLayer()
            }

            is FillLayer -> {
                val fillLayer = layer
                sourceId = fillLayer.getSourceId()
                sourceLayerId = fillLayer.getSourceLayer()
            }

            is HeatmapLayer -> {
                val heatmapLayer = layer
                sourceId = heatmapLayer.getSourceId()
                sourceLayerId = heatmapLayer.getSourceLayer()
            }

            is HillshadeLayer -> {
                val hillshadeLayer = layer
                sourceId = hillshadeLayer.getSourceId()
                sourceLayerId = null
            }

            is LineLayer -> {
                val lineLayer = layer
                sourceId = lineLayer.getSourceId()
                sourceLayerId = lineLayer.getSourceLayer()
            }

            is RasterLayer -> {
                val rasterLayer = layer
                sourceId = rasterLayer.getSourceId()
                sourceLayerId = null
            }

            is SymbolLayer -> {
                val symbolLayer = layer
                sourceId = symbolLayer.getSourceId()
                sourceLayerId = symbolLayer.getSourceLayer()
            }

            else -> {
                sourceId = ""
                sourceLayerId = null
            }
        }
    }
}
